drivers: spi: Add simple_spi driver

olofk opened the pull request:
Add support for the OpenCores simple_spi controller

This driver is used in the [SweRVolf](https://github.com/chipsalliance/Cores-SweRVolf) SoC and has been tested on a Nexys A7 board with the Zephyr SPI Flash sample as well as in simulation

Signed-off-by: Olof Kindgren <olof.kindgren@gmail.com>

olofk committed:
drivers: spi: Add simple_spi driver

Add support for the OpenCores simple_spi controller

Signed-off-by: Olof Kindgren <olof.kindgren@gmail.com>

tbursztyka commented on drivers/spi/spi_simple_spi.c:
```diff
@@ -0,0 +1,177 @@
+	/* Set chip select */
+	if (config->cs == NULL) {
+		sys_write8(1 << config->slave, SPI_REG(dev, SPI_SPSS));
+	}
```
please, support as well the possibility to emulate the cs line via a gpio line. See other drivers, calling spi_context_cs_configure() (then spi_context_cs_control() )

olofk replied:
Done

galak reviewed (changes requested):
Can we rename the files and Kconfig options to have opencores in the name.  simple_spi by itself can be a little confusing/misleading.

olofk commented on drivers/spi/spi_simple_spi.c:
```diff
@@ -0,0 +1,177 @@
+	spi_context_lock(ctx, false, NULL);
+
+	/* Set clock divider */
+	int i;
```
Done

olofk commented on drivers/spi/spi_simple_spi.c:
```diff
@@ -0,0 +1,177 @@
+				sys_write8(0, SPI_REG(dev, SPI_SPDR));
+			}
+
+			__asm__ volatile ("fence;\n\t");
```
Removed. It was a leftover from some earlier bug hunting

olofk commented on drivers/spi/spi_simple_spi.c:
```diff
@@ -0,0 +1,177 @@
+static struct spi_driver_api simple_spi_api = {
+	.transceive = simple_spi_transceive,
+	.release = simple_spi_release,
+};
```
Done

olofk commented on drivers/spi/spi_simple_spi.c:
```diff
@@ -0,0 +1,177 @@
+
+	sys_write8((DIVIDERS[i] >> 4) & 0x3, SPI_REG(dev, SPI_SPER));
+	u8_t spcr = DIVIDERS[i] & 0x3;
+
```
Done. Also expanded the configure function to handle all unsupported features

olofk commented on drivers/spi/spi_simple_spi.h:
```diff
@@ -0,0 +1,29 @@
+
+#include "spi_context.h"
+
+#define SPI_CFG(dev)  ((struct simple_spi_cfg  *) ((dev)->config->config_info))
```
Done. I ended up with a whole new set of macros that make more sense

olofk commented:
Thanks for the review. I have corrected all the issues that @tbursztyka found. I'm struggling a bit with the naming though. The core has always been referred to as Simple SPI. The opencores,spi-simple device tree entry was named to keep it compatible with the [Linux driver](https://github.com/skristiansson/linux/blob/openrisc/drivers/spi/spi-oc-simple.c#L456). I guess we want to have the spi_ prefix for all files in the spi drivers directory. I renamed it to `spi_oc_simple` now. There are more than ten different SPI controllers on OpenCores so I don't want to just call it spi_opencores or spi_oc. Should I just force-push a new version when I'm done?

galak commented:
> Thanks for the review. I have corrected all the issues that @tbursztyka found. I'm struggling a bit with the naming though that @galak commented on. The core has always been referred to as Simple SPI. The opencores,spi-simple device tree entry was named to keep it compatible with the [Linux driver](https://github.com/skristiansson/linux/blob/openrisc/drivers/spi/spi-oc-simple.c#L456). I guess we want to have the spi_ prefix for all files in the spi drivers directory. I renamed it to `spi_oc_simple` now. There are more than ten different SPI controllers on OpenCores so I don't want to just call it spi_opencores or spi_oc. Should I just force-push a new version when I'm done?

`spi_oc_simple` is fine.  Yes just force push and update when you've made the changes.

olofk commented:
Thanks @galak New one is going through CI checks now
